README.md
# ntp-gps-tools
Bunch of hacks to setup anRPi based gps driven NTP Stratum 1 server


gpsctl/Changelog
2020-10-20 - Phil Randal - Release 1.13
    Added After=network-online.target to ublox-init.service
    This gives me reliable baud rate autodetection at startup
    Minor patch to cJSON.c, not needed by us but who knows who borrows what from where

2020-10-18 - Phil Randal - Release 1.12
    Increased --autobaud synchronisation timeout from 3 to 4 seconds

2020-10-18 - Phil Randal - Release 1.11
    Increased --autobaud synchronisation timeout from 1.5 to 3 seconds

2020-10-18 - Phil Randal - Release 1.10b
    Added configurable parameters to gpsctl in ublox-init.service
    Set parameters in /etc/default/gpstl, e.g.
        PARAMS=-q --configure_for_timing

2020-09-07 - Phil Randal - Release 1.10
    Added gpsctl.conf.ublox which matches the ublox config after gpsctl --reset
    Added NMEA GGA config to gpsctl -Q config output

2020-09-03 - Phil Randal
    Update to cJSON 1.7.14

2020-07-17 - Phil Randal
    Fix possible null pointer dereference in iniparser.c

2020-06-21 - Phil Randal - Release 1.9
    Set configure-rv3028.sh as executable
    Reduce number of characters from 50 to 25 in baud rate autodetection

2020-06-21 - Phil Randal - Release 1.8
    Run stty in ublox-init.service to fix startup glitches
    Add configure-rv3028.sh to configure the RTC on newer Uputronics GPS Hats

2020-05-16 - Phil Randal
    Update to cJSON 1.7.13

2019-07-20 - Phil Randal - Release 1.7
    Allow BeiDou and Glonass to co-exist as long as neither GPS nor Galileo are enabled

2019-07-20 - Phil Randal - Release 1.6
    Add "prefer beidou to glonass if both enabled" option in
       [gpsctl] config file section.  Defaults to 'no'.
       So if you accidentally try to enable both, you can
       specify which one of the two will be enabled

2019-07-15 - Phil Randal - Release 1.5
    Enable NMEA 4.1 for both Beidou and Galileo

2019-05-25 - Phil Randal - Release 1.4
    Changed default gpsctl.conf to match original program's apart from
        C/No threshold reduced to 20 dBHz

2019-05-24 - Phil Randal
    Reformat gpsctl.c and ublox.c to get rid of irritating whitespace
    Update to cJSON 1.7.12
    Added MIT LICENSE file

2019-05-07 - Phil Randal - Release 1.3
    Full reporting of enabled / disabled NMEA messages
    Default to NMEA versioning in x.y format

2019-05-07 - Phil Randal
    Minor bug fixes to iniparser
    Turn off debug symbols in gpsctl executable

2019-05-06 - Phil Randal
    Minor change to NMEA version reporting to make it consistent

2019-05-01 - Phil Randal - Release 1.2
    Fix iniparser config dump code to correctly quote values containing ';' and '#'
    Allow non boolean option for nmea:enabled to leave current setting undisturbed

2019-05-01 - Phil Randal - Release 1.1

2019-04-29 - Phil Randal
    Turn xxRMC on by default in the example gpsctl.conf file

2019-04-28 - Phil Randal
    Replace ciniparser with iniparser (from my own fork)

2019-04-26 - Phil Randal - Release 1.0
    Add defaults for "verbosity" and "sync method" to gpsctl.conf
    Both can be overridden on the command line
    Add nmea enabled option to config file

2019-04-26 - Phil Randal - Release 0.9:
    Add --satellites option (same as --galileo, which is now deprecated)
    Force NMEA version 41 or higher if Galileo enabled

2019-04-26 - Phil Randal - Release 0.8:
    Add enabling / disabling NMEA sentences, timing and navigation engine config parameters to gpsctl.conf

2019-04-25 - Phil Randal - Release 0.7:
    Add /etc/gpsctl.conf file supporting satellite and nmea version configuration

2019-04-23 - Phil Randal
    Add ciniparser into the mix for future gpsctl.conf file parsing

2019-04-18 - Phil Randal
    Refactor some of the code
    Report Fix Date and time in ISO 8601 format (yyyy-mm-dd hh:mm:ss)

2019-04-17 - Phil Randal - Release 0.6:
    Updated cJSON from version 1.6.0 to 1.7.11

2019-04-15 - Phil Randal:
    Added --galileo option to enable Galileo Satellites and NMEA 4.1 separately (implied by --configure_for_timing)
    Relaxed satellite fix constraints so I could actually get a fix using active GPS antenna
    Disabled NMEA RMC, VTG, GSA, GSV, GLL, GGA and enabled ZDA for use with NTP driver 20 option 8
    Fixed parsing of --port option to actually pass the specified port parameter
    Changed version number to 0.6


gpsctl/sl_general.h
//
// Created by Tom Dilatush on 10/19/17.
//

#ifndef GPSCTL_SLIGHTLY_LOONY_H
#define GPSCTL_SLIGHTLY_LOONY_H
#include <memory.h>
#include <stdlib.h>
#include <execinfo.h>
#include <stdio.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdarg.h>
#include <stdint.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/stat.h>
#include <time.h>
#include <errno.h>
#include <ctype.h>
#
#define ARRAY_SIZE(x) (sizeof (x) / sizeof *(x))

typedef enum Endiness { BigEndian, LittleEndian } Endiness;

typedef unsigned char byte;

char* onOff( const bool onoff );

char* yesNo( const bool yesno );

void stackDump( void );

void *safeMalloc( size_t size );

bool issprint( const char* str );

bool issgraph( const char* str );

char* concat( const char *s1, const char *s2 );

void append( char** s1, const char* s2 );

void sleep_ns( int ns );

long long currentTimeMs();

bool strempty( const char* str );

long long max_ll( long long a, long long b );
int min_i( int a, int b);

int hex2int( char c );

bool isLeapYear( uint32_t year );

uint32_t daysInMonth( uint32_t year, uint32_t month );

char* getAllocatedStringCopy( const char *source );

// Returns the number of arguments that have been passed to it.
#define NUMARGS(...) NUMARGS_(__VA_ARGS__,RSEQ_N())
#define NUMARGS_(...) NUMARGS_N(__VA_ARGS__)
#define NUMARGS_N( \
          _1, _2, _3, _4, _5, _6, _7, _8, _9,_10, \
         _11,_12,_13,_14,_15,_16,_17,_18,_19,_20, \
         _21,_22,_23,_24,_25,_26,_27,_28,_29,_30, \
         _31,_32,_33,_34,_35,_36,_37,_38,_39,_40, \
         _41,_42,_43,_44,_45,_46,_47,_48,_49,_50, \
         _51,_52,_53,_54,_55,_56,_57,_58,_59,_60, \
         _61,_62,_63,N,...) N
#define RSEQ_N() \
         63,62,61,60,                   \
         59,58,57,56,55,54,53,52,51,50, \
         49,48,47,46,45,44,43,42,41,40, \
         39,38,37,36,35,34,33,32,31,30, \
         29,28,27,26,25,24,23,22,21,20, \
         19,18,17,16,15,14,13,12,11,10, \
         9,8,7,6,5,4,3,2,1,0

#endif //GPSCTL_SLIGHTLY_LOONY_H
